SpatialGDK/Source/Private/Interop/GlobalStateManager.cpp

```diff
@@ -93,170 +93,168 @@ void UGlobalStateManager::ApplyAcceptingPlayersUpdate(bool bAcceptingPlayersUpda
+	if (Channel != nullptr)
+	{
+		// Channel has already been setup
+		UE_LOG(LogGlobalStateManager, Warning, TEXT("UGlobalStateManager::LinkExistingSingletonActor channel already setup"), *SingletonActorClass->GetName());
+		return;
 	}
-}
 
-void UGlobalStateManager::ExecuteInitialSingletonActorReplication()
-{
-	for (const auto& Pair : SingletonNameToEntityId)
-	{
-		Worker_EntityId SingletonEntityId = Pair.Value;
+	// If we have previously queued up data for this entity, apply it - JIRA: 734
 
-		// Entity has already been created on another server
-		if (SingletonEntityId != SpatialConstants::INVALID_ENTITY_ID)
-		{
-			continue;
-		}
+	USpatialNetConnection* Connection = Cast<USpatialNetConnection>(NetDriver->ClientConnections[0]);
+	Channel = Cast<USpatialActorChannel>(Connection->CreateChannel(CHTYPE_Actor, 1));
 
-		AActor* SingletonActor = nullptr;
-		USpatialActorChannel* Channel = nullptr;
-		GetSingletonActorAndChannel(Pair.Key, SingletonActor, Channel);
+	SingletonActor->Role = ROLE_SimulatedProxy;
+	SingletonActor->RemoteRole = ROLE_Authority;
 
-		// Class couldn't be found or channel already exists
-		if (Channel == nullptr || Channel->Actor != nullptr)
-		{
-			check(Channel->Actor == SingletonActor);
-			continue;
-		}
+	// Add to entity registry
+	// This indirectly causes SetChannelActor to not create a new entity for this actor
+	NetDriver->GetEntityRegistry()->AddToRegistry(SingletonEntityId, SingletonActor);
 
-		if (!SingletonActor->GetIsReplicated())
-		{
-			UE_LOG(LogGlobalStateManager, Warning, TEXT("Singleton Actor %s isn't replicated! No entity will be created."), *SingletonActor->GetClass()->GetName());
-			continue;
-		}
+	Channel->SetChannelActor(SingletonActor);
 
-		SingletonActor->Role = ROLE_Authority;
-		SingletonActor->RemoteRole = ROLE_SimulatedProxy;
+	// Since the entity already exists, we have to handle setting up the PackageMap properly for this Actor
+	FClassInfo* Info = NetDriver->TypebindingManager->FindClassInfoByClass(SingletonActor->GetClass());
+	check(Info);
 
-		// Set entity id of channel from the GlobalStateManager.
-		// If the id was 0, SetChannelActor will create the entity.
-		// If the id is not 0, it will start replicating to that entity.
-		Channel->SetChannelActor(SingletonActor);
+	NetDriver->PackageMap->ResolveEntityActor(SingletonActor, SingletonEntityId, improbable::CreateOffsetMapFromActor(SingletonActor, Info));
 
-		UE_LOG(LogGlobalStateManager, Log, TEXT("Started replication of Singleton Actor %s"), *SingletonActor->GetClass()->GetName());
-	}
+	UE_LOG(LogGlobalStateManager, Log, TEXT("Linked Singleton Actor %s with id %d"), *SingletonActor->GetClass()->GetName(), SingletonEntityId);
 }
 
-void UGlobalStateManager::UpdateSingletonEntityId(const FString& ClassName, const Worker_EntityId SingletonEntityId)
+void UGlobalStateManager::LinkExistingSingletonActors()
```

Comment:
Maybe we should call this `LinkAllExistingSingletonActors` just to make it more distinct.